File: src/main/java/io/github/chaosdave34/kitpvp/damagetype/DamageTypes.java
package io.github.chaosdave34.kitpvp.damagetype;

import io.github.chaosdave34.ghutils.utils.ReflectionUtils;
import net.minecraft.core.Registry;
import net.minecraft.core.registries.Registries;
import net.minecraft.world.damagesource.DamageEffects;
import net.minecraft.world.damagesource.DamageScaling;
import org.bukkit.NamespacedKey;
import org.bukkit.craftbukkit.v1_20_R3.CraftRegistry;
import org.bukkit.craftbukkit.v1_20_R3.damage.CraftDamageType;
import org.bukkit.damage.DamageType;
import org.jetbrains.annotations.ApiStatus;

@ApiStatus.Experimental
public class DamageTypes {
    private final Registry<net.minecraft.world.damagesource.DamageType> registry;

    public static DamageType LAND;
    public static DamageType ESCAPE;

    public DamageTypes() {
        registry = CraftRegistry.getMinecraftRegistry(Registries.DAMAGE_TYPE);
        ReflectionUtils.unfreezeRegistry(registry);

        LAND = registerDamageType(new net.minecraft.world.damagesource.DamageType("land", DamageScaling.NEVER, 0, DamageEffects.HURT), "land");
        ESCAPE = registerDamageType(new net.minecraft.world.damagesource.DamageType("escape", DamageScaling.NEVER, 0, DamageEffects.HURT), "escape");
    }

    private DamageType registerDamageType(net.minecraft.world.damagesource.DamageType damageType, String id) {
        registry.createIntrusiveHolder(damageType);
        Registry.register(registry, NamespacedKey.minecraft(id).asString(), damageType);

        return CraftDamageType.minecraftToBukkit(damageType);
    }
}

File: src/main/kotlin/io/github/chaosdave34/kitpvp/damagetype/DamageTypes.kt
package io.github.chaosdave34.kitpvp.damagetype

import io.github.chaosdave34.kitpvp.KitPvp
import net.minecraft.core.MappedRegistry
import net.minecraft.core.Registry
import net.minecraft.core.registries.Registries
import net.minecraft.world.damagesource.DamageEffects
import net.minecraft.world.damagesource.DamageScaling
import org.bukkit.NamespacedKey
import org.bukkit.craftbukkit.CraftRegistry
import org.bukkit.craftbukkit.damage.CraftDamageType
import org.bukkit.damage.DamageType
import java.util.*
import net.minecraft.world.damagesource.DamageType as NmsDamageType

class DamageTypes {
    private var registry : Registry<NmsDamageType> = CraftRegistry.getMinecraftRegistry(Registries.DAMAGE_TYPE)

    companion object {
        lateinit var LAND: DamageType
        lateinit var ESCAPE: DamageType
        lateinit var BLACK_HOLE: DamageType
    }

    init {
        // Unfreeze registry HIGHLY UNSUPPORTED!!!
        try {
            for (field in MappedRegistry::class.java.declaredFields) {
                // private Map<T, Holder.Reference<T>> unregisteredIntrusiveHolders
                if (field.name == "unregisteredIntrusiveHolders" || field.name == "m") {
                    field.isAccessible = true
                    field[registry] = IdentityHashMap<Any, Any>()
                    field.isAccessible = false
                } else if (field.name == "frozen" || field.name == "l") {
                    field.isAccessible = true
                    field[registry] = false
                    field.isAccessible = false
                }
            }
        } catch (e: IllegalAccessException) {
            KitPvp.INSTANCE.logger.warning("Error while unfreezing damageType registry.")
        }

        LAND = registerDamageType(NmsDamageType("land", DamageScaling.NEVER, 0f, DamageEffects.HURT))
        ESCAPE = registerDamageType(NmsDamageType("escape", DamageScaling.WHEN_CAUSED_BY_LIVING_NON_PLAYER, 0f, DamageEffects.HURT))
        BLACK_HOLE = registerDamageType(NmsDamageType("black_hole", DamageScaling.NEVER, 0f, DamageEffects.HURT))
    }

    private fun registerDamageType(damageType: NmsDamageType): DamageType {
        registry.createIntrusiveHolder(damageType)
        Registry.register(registry, NamespacedKey.minecraft(damageType.msgId).asString(), damageType)

        return CraftDamageType.minecraftToBukkit(damageType)
    }
}